tensorflow_quantum/core/ops/tfq_resolve_parameters_op.cc

```diff
@@ -0,0 +1,106 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <string>
+
+#include "cirq/google/api/v2/program.pb.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/lib/core/error_codes.pb.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/core/threadpool.h"
+#include "tensorflow_quantum/core/ops/parse_context.h"
+#include "tensorflow_quantum/core/ops/tfq_simulate_utils.h"
+#include "tensorflow_quantum/core/src/program_resolution.h"
+
+namespace tfq {
+
+using ::cirq::google::api::v2::Program;
+using ::tensorflow::Status;
+
+class TfqResolveParametersOp : public tensorflow::OpKernel {
+ public:
+  explicit TfqResolveParametersOp(tensorflow::OpKernelConstruction* context)
+      : OpKernel(context) {}
+
+  void Compute(tensorflow::OpKernelContext* context) override {
+    const int num_inputs = context->num_inputs();
+    OP_REQUIRES(context, num_inputs == 3,
+                tensorflow::errors::InvalidArgument(absl::StrCat(
+                    "Expected 3 inputs, got ", num_inputs, " inputs.")));
+
+    tensorflow::Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(
+                                0, context->input(0).shape(), &output));
+    auto output_tensor = output->flat<tensorflow::tstring>();
+
+    std::vector<Program> programs;
+    OP_REQUIRES_OK(context, ParsePrograms(context, "programs", &programs));
+    std::vector<SymbolMap> maps;
+    OP_REQUIRES_OK(context, GetSymbolMaps(context, &maps));
+
+    OP_REQUIRES(
+        context, maps.size() == programs.size(),
+        tensorflow::errors::InvalidArgument(absl::StrCat(
+            "Number of circuits and values do not match. Got ", programs.size(),
+            " circuits and ", maps.size(), " values.")));
+
+    auto DoWork = [&](int start, int end) {
+      std::string temp;
+      for (int i = start; i < end; i++) {
+        Program program = programs[i];
+        OP_REQUIRES_OK(context, ResolveSymbols(maps[i], &program, false));
+        program.SerializeToString(&temp);
+        output_tensor(i) = temp;
+      }
+    };
+
+    const int output_dim_size = programs.size();
+    const int block_size = GetBlockSize(context, output_dim_size);
+    context->device()
+        ->tensorflow_cpu_worker_threads()
+        ->workers->TransformRangeConcurrently(block_size, output_dim_size,
+                                              DoWork);
```

Comment:
We want to try and use `ParallelFor` here instead of `TransformRangeConcurrently`.

Reply:
Changed to `ParallelFor`